File: backend/app/schemas/voucher.py
"""记账凭证 Pydantic 模型。"""
from typing import List, Optional

from pydantic import BaseModel


class VoucherEntryRead(BaseModel):
    seq: int
    subject_code: str
    subject_name: str
    summary: Optional[str] = None
    direction: str   # 借/贷
    amount: float


class VoucherEntryCreate(BaseModel):
    """手工凭证分录（写入用）。"""
    subject_code: str
    summary: Optional[str] = None
    direction: str                     # 借/贷
    amount: float


class VoucherCreate(BaseModel):
    """手工录入凭证（开放手工凭证入口，独立 source_type='手工' 不污染业务联动）。"""
    voucher_date: str                     # YYYY-MM-DD
    voucher_word: str = "记"             # 记/收/付/转
    attach_count: int = 0
    maker: Optional[str] = None
    summary: Optional[str] = None
    entries: List[VoucherEntryCreate]


class VoucherRead(BaseModel):
    id: int
    voucher_no: str
    date: str           # ISO 日期字符串
    period: str          # 会计期间：2026-07
    voucher_word: str    # 凭证字：记/收/付/转
    seq: int             # 凭证序号（凭证号末段）
    attach_count: int
    maker: Optional[str] = None
    status: str          # 未审核/已审核/已记账
    source_type: Optional[str] = None
    source_no: Optional[str] = None
    summary: Optional[str] = None
    entries: List[VoucherEntryRead] = []


class VoucherGenerateResult(BaseModel):
    """批量从已通过业务单补生成凭证的结果。"""

    generated: int
    skipped: int
    detail: List[str] = []

File: frontend/src/api/voucher.ts
/** 记账凭证 API 客户端：状态机（审核/记账/反审核/反记账）+ 手工录入。 */
import request from '@/utils/request'

/** 手工凭证分录（写入用） */
export interface VoucherEntryCreate {
  subject_code: string
  summary?: string
  direction: '借' | '贷'
  amount: number
}

/** 手工录入凭证（写入用） */
export interface VoucherCreate {
  voucher_date: string            // YYYY-MM-DD
  voucher_word?: string        // 记/收/付/转
  attach_count?: number
  maker?: string
  summary?: string
  entries: VoucherEntryCreate[]
}

/**
 * 手工录入凭证：独立 source_type='手工'，借贷平衡 + 至少 2 条分录。
 * 对应后端 POST /api/vouchers
 */
export function createVoucher(payload: VoucherCreate) {
  return request.post('/vouchers', payload)
}

/**
 * 审核凭证：未审核 → 已审核（已审核/已记账幂等返回当前态）。
 * 对应后端 POST /api/vouchers/{id}/audit
 */
export function auditVoucher(id: number) {
  return request.post(`/vouchers/${id}/audit`)
}

/**
 * 记账凭证：已审核 → 已记账（未审核不可直接记账，后端返 400）。
 * 对应后端 POST /api/vouchers/{id}/post
 */
export function postVoucher(id: number) {
  return request.post(`/vouchers/${id}/post`)
}

/**
 * 反记账：已记账 → 已审核（释放记账锁定，仍在账簿中）。
 * 对应后端 POST /api/vouchers/{id}/unpost
 */
export function unpostVoucher(id: number) {
  return request.post(`/vouchers/${id}/unpost`)
}

/**
 * 反审核：已审核 → 未审核（拉出账簿）；已记账须先反记账。
 * 对应后端 POST /api/vouchers/{id}/unaudit
 */
export function unauditVoucher(id: number) {
  return request.post(`/vouchers/${id}/unaudit`)
}
